pkg/config/expression.go

```diff
@@ -64,47 +67,216 @@ func SimpleVarToReference(s string) (Reference, error) {
 		if components[0] == "vars" {
 			return Reference{
 				GlobalVar: true,
-				Name:      components[1]}, nil
+				Name:      components[1]}, "", nil
 		}
 		return Reference{
 			Module: components[0],
-			Name:   components[1]}, nil
+			Name:   components[1]}, "", nil
 
 	case 3:
 		return Reference{
-			Group:  components[0],
 			Module: components[1],
-			Name:   components[2]}, nil
+			Name:   components[2]}, components[0], nil
 	default:
-		return Reference{}, fmt.Errorf(
+		return Reference{}, "", fmt.Errorf(
 			"expected either 2 or 3 components, got %d in %#v", len(components), s)
 	}
 }
 
-// VariableTranslator is an interface that provides function
-// to translate "simple" variable (`$(...)`) into HCL format
-type VariableTranslator interface {
-	TranslateSimpleToHcl(s string) (string, error)
+// SimpleVarToHclExpression takes a string `$(...)` and transforms it to `HclExpression`
+// Returns `HclExpression` and name of explicit group if one was set.
+func SimpleVarToHclExpression(s string) (HclExpression, string, error) {
+	ref, gr, err := SimpleVarToReference(s)
+	if err != nil {
+		return HclExpression{}, "", err
+	}
+	var ex HclExpression
+	if ref.GlobalVar {
+		ex, err = ParseExpression(fmt.Sprintf("var.%s", ref.Name))
+	} else {
+		ex, err = ParseExpression(fmt.Sprintf("module.%s.%s", ref.Module, ref.Name))
+	}
+	if err != nil {
+		return HclExpression{}, "", err
+	}
+	return ex, gr, nil
+}
+
+// TraversalToReference takes HCL traversal and returns `Reference`
+func TraversalToReference(t hcl.Traversal) (Reference, error) {
+	if t.IsRelative() {
+		return Reference{}, fmt.Errorf("got relative traversal")
+	}
+	getAttrName := func(i int) (string, error) {
+		if i >= len(t) {
+			return "", fmt.Errorf("traversal does not have enough components")
+		}
+		if a, ok := t[i].(hcl.TraverseAttr); ok {
```

Comment:
what cases do we expect this cast to fail and error "got unexpected..."? Any realistic ones?

Reply:
Any of those: `var`, `module`, `module.module_name`